<a href="https://colab.research.google.com/github/priyansh1997/AlmaBetter/blob/main/Priyansh_Tripathi_EDA_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1.1 Importing Libraries
These libraries will be used for Exploratory Data Analysis.

In [96]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [97]:
# Removing warnings
import warnings
warnings.filterwarnings("ignore")

# 1.2 Data Loading
Loading the dataset from google drive.

In [98]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [99]:
play_store_df=pd.read_csv('/content/drive/MyDrive/datasets/Play Store Data.csv')
user_reviews_df=pd.read_csv('/content/drive/MyDrive/datasets/User Reviews.csv')

# 1.3 Data Inspection
Inspection of the data contains information about the data, head/tail part, description of dataset, shape etc.

In [100]:
play_store_df.head()

App  ...   Android Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook  ...  4.0.3 and up
1                                Coloring book moana  ...  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ...  4.0.3 and up
3                              Sketch - Draw & Paint  ...    4.2 and up
4              Pixel Draw - Number Art Coloring Book  ...    4.4 and up

[5 rows x 13 columns]

In [101]:
play_store_df.tail()

App  ...         Android Ver
10836                               Sya9a Maroc - FR  ...          4.1 and up
10837               Fr. Mike Schmitz Audio Teachings  ...          4.1 and up
10838                         Parkinson Exercices FR  ...          2.2 and up
10839                  The SCP Foundation DB fr nn5n  ...  Varies with device
10840  iHoroscope - 2018 Daily Horoscope & Astrology  ...  Varies with device

[5 rows x 13 columns]

In [102]:
user_reviews_df.head()

App  ... Sentiment_Subjectivity
0  10 Best Foods for You  ...               0.533333
1  10 Best Foods for You  ...               0.288462
2  10 Best Foods for You  ...                    NaN
3  10 Best Foods for You  ...               0.875000
4  10 Best Foods for You  ...               0.300000

[5 rows x 5 columns]

In [103]:
user_reviews_df.tail()

App  ... Sentiment_Subjectivity
64290  Houzz Interior Design Ideas  ...                    NaN
64291  Houzz Interior Design Ideas  ...                    NaN
64292  Houzz Interior Design Ideas  ...                    NaN
64293  Houzz Interior Design Ideas  ...                    NaN
64294  Houzz Interior Design Ideas  ...                    NaN

[5 rows x 5 columns]

In [104]:
play_store_df.shape

(10841, 13)

In [105]:
user_reviews_df.shape

(64295, 5)

In [106]:
play_store_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


Here we are having table showing 13 feature names and the count of non-values present in them. Now we will check each column in data cleaning part. 

In [107]:
user_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


Here also in user reviews dataframe we have 5 features with there counts of non-null values.

In [108]:
play_store_df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [109]:
# Checking the column names and removing the spaces so column will be easier to access 
play_store_df.rename(columns={"Content Rating":"Content_Rating", "Last Updated":"Last_Updated", "Current Ver":"Current_Ver", "Android Ver":"Android_Ver"}, inplace=True)

In [110]:
play_store_df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content_Rating', 'Genres', 'Last_Updated', 'Current_Ver',
       'Android_Ver'],
      dtype='object')

In [111]:
user_reviews_df.columns 

Index(['App', 'Translated_Review', 'Sentiment', 'Sentiment_Polarity',
       'Sentiment_Subjectivity'],
      dtype='object')

Already spaces are replaced by underscores

### 1.3.1 Data Cleaning part for Play store dataframe.

Firstly we will remove the duplicated rows for the play store dataframe.

In [112]:
duplicated_rows_play_store_index=play_store_df[play_store_df.duplicated()].index
print(duplicated_rows_play_store_index)

Int64Index([  229,   236,   239,   256,   261,   265,   266,   267,   268,
              269,
            ...
             8632,  8635,  8636,  8639,  8642,  8643,  8654,  8658, 10049,
            10768],
           dtype='int64', length=483)


In [113]:
# Dropping the duplicated rows as they were disturbing the data integrity and checking the shape
play_store_df.drop(duplicated_rows_play_store_index, inplace=True)
play_store_df.shape

(10358, 13)

In [114]:
play_store_df.App.isnull().sum()

0

No null values in App feature.

In [115]:
play_store_df.Category.isnull().sum()

0

No null values in Category feature.

In [116]:
# Getting the unique values of category column
play_store_df.Category.unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

These are all the unique categories of application that we are having in our data.

In [117]:
play_store_df.Rating.isnull().sum()

1465

We are having 1465 null values in rating column which we have to fill or drop out.

In [118]:
print("Minimum rating to an App given:",min(play_store_df.Rating))
print("Maximum rating to an App given:",max(play_store_df.Rating))

Minimum rating to an App given: 1.0
Maximum rating to an App given: 19.0


Max rating of the App can be 5.0

In [119]:
play_store_df[play_store_df.Rating>5]

App Category  ...  Current_Ver Android_Ver
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9  ...   4.0 and up         NaN

[1 rows x 13 columns]

We can see this record is manipulated as it has rating more than 5, price is everyone which should be some value, installs is free etc. so we will remove this record.

In [120]:
play_store_df.drop(play_store_df[play_store_df.Rating>5].index,inplace=True)

In [121]:
print("Now maximum rating of an App is:",max(play_store_df.Rating))

Now maximum rating of an App is: 5.0


In [122]:
play_store_df.reset_index(drop=True,inplace=True)

In [123]:
play_store_df.head()

App  ...   Android_Ver
0     Photo Editor & Candy Camera & Grid & ScrapBook  ...  4.0.3 and up
1                                Coloring book moana  ...  4.0.3 and up
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ...  4.0.3 and up
3                              Sketch - Draw & Paint  ...    4.2 and up
4              Pixel Draw - Number Art Coloring Book  ...    4.4 and up

[5 rows x 13 columns]

In [124]:
play_store_df.Rating.isnull().sum()

1465

Still we are having 1465 number of null values in our Rating column.

In [125]:
versions_of_app=play_store_df[play_store_df.App.duplicated(keep=False)].groupby(by=["App"],as_index=False).apply(lambda x: x[:])
versions_of_app

App  ...         Android_Ver
0   2697                            365Scores - Live Scores  ...          4.1 and up
    4959                            365Scores - Live Scores  ...          4.1 and up
1   1468                                        8 Ball Pool  ...        4.0.3 and up
    1496                                        8 Ball Pool  ...        4.0.3 and up
    1544                                        8 Ball Pool  ...        4.0.3 and up
...                                                     ...  ...                 ...
520 2331                        textPlus: Free Text & Calls  ...          4.1 and up
521 2602  theScore: Live Sports Scores, News, Stats & Vi...  ...          4.4 and up
    2680  theScore: Live Sports Scores, News, Stats & Vi...  ...          4.4 and up
522 3226                     wetter.com - Weather and Radar  ...  Varies with device
    7820                     wetter.com - Weather and Radar  ...  Varies with device

[1221 rows x 13 columns]

In [126]:
versions_of_app[versions_of_app["Rating"].isnull()]

App        Category  ...  Current_Ver Android_Ver
48  2222          Blood Pressure         MEDICAL  ...          1.0  2.3 and up
282 23    Mcqueen Coloring pages  ART_AND_DESIGN  ...        1.0.0  4.1 and up
    1878  Mcqueen Coloring pages          FAMILY  ...        1.0.0  4.1 and up

[3 rows x 13 columns]

Checking for the different versions and filling the rating of Applications with there mean.

In [127]:
play_store_df[play_store_df["App"]=='Blood Pressure']

App Category  Rating  ...    Last_Updated Current_Ver Android_Ver
2052  Blood Pressure  MEDICAL     4.2  ...   July 24, 2018      3.27.3  4.1 and up
2222  Blood Pressure  MEDICAL     NaN  ...  March 14, 2015         1.0  2.3 and up

[2 rows x 13 columns]

Only 2 versions available so we can fill with the mean of category to which this app belongs.

In [128]:
play_store_df[play_store_df["App"]=='Mcqueen Coloring pages']

App        Category  ...  Current_Ver Android_Ver
23    Mcqueen Coloring pages  ART_AND_DESIGN  ...        1.0.0  4.1 and up
1878  Mcqueen Coloring pages          FAMILY  ...        1.0.0  4.1 and up

[2 rows x 13 columns]

Here also only 2 applications are there and they are also present in the different category so we will put catgory wise mean in both.

In rest of the records with null-values in rating column we will fill category wise mean because we don't have different versions of those applications.

In [129]:
play_store_df['Rating'].fillna(play_store_df.groupby('Category')['Rating'].transform('mean'), inplace=True)

In [130]:
# Checking the Reivew column
play_store_df['Reviews']=play_store_df['Reviews'].astype('int')

In [131]:
play_store_df['Size'].unique()[30:70]

array(['9.4M', '15M', '10M', '1.2M', '26M', '8.0M', '7.9M', '56M', '57M',
       '35M', '54M', '201k', '3.6M', '5.7M', '8.6M', '2.4M', '27M',
       '2.5M', '16M', '3.4M', '8.9M', '3.9M', '2.9M', '38M', '32M',
       '5.4M', '18M', '1.1M', '2.2M', '4.5M', '9.8M', '52M', '9.0M',
       '6.7M', '30M', '2.6M', '7.1M', '3.7M', '22M', '7.4M'], dtype=object)

As we can see we have size of the app in Mb's and Kb's we have to convert these into same unit and also remove the unwanted symbols.

In [132]:
play_store_df['Size'] = play_store_df['Size'].apply(lambda x: str(x).replace('Varies with device', 'NaN') if 'Varies with device' in str(x) else x)
play_store_df['Size'] = play_store_df['Size'].apply(lambda x: str(x).replace('M', '') if 'M' in str(x) else x)
play_store_df['Size'] = play_store_df['Size'].apply(lambda x: str(x).replace(',', '') if ',' in str(x) else x)
play_store_df['Size'] = play_store_df['Size'].apply(lambda x: float(str(x).replace('k', '')) / 1024 if 'k' in str(x) else x)
play_store_df['Size'] = play_store_df['Size'].apply(lambda x: str(x).replace('+','') if '+' in str(x) else x)

In [133]:
play_store_df['Size'].dtype

dtype('O')

In [134]:
# Type Casting the Size feature in float64 
play_store_df.Size=play_store_df.Size.astype('float64')

In [135]:
# Checking the data type
play_store_df['Size'].dtype

dtype('float64')

In [136]:
play_store_df['Size'].isnull().sum()

1526

We have 1526 null values filling them with category wise mean of the Size.

In [137]:
play_store_df['Size'].fillna(play_store_df.groupby('Category')['Size'].transform('mean'), inplace=True)

In [138]:
# Checking for the Installs feature
play_store_df.Installs.unique(),len(play_store_df.Installs.unique())

(array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
        '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
        '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
        '10+', '1+', '5+', '0+', '0'], dtype=object), 21)

We can use the Installs columns category wise too and numerical wise too in our EDA section so we will not change it over here.

In [139]:
play_store_df.Type.unique()

array(['Free', 'Paid', nan], dtype=object)

In [140]:
# Type feature null values
play_store_df[play_store_df.Type.isnull()]

App  ...         Android_Ver
8667  Command & Conquer: Rivals  ...  Varies with device

[1 rows x 13 columns]

We have only single value so we will fill the Type in this record manually.

In [141]:
play_store_df['Price'].unique()[5:15]

array(['$2.99', '$7.99', '$5.99', '$3.49', '$1.99', '$9.99', '$7.49',
       '$0.99', '$9.00', '$5.49'], dtype=object)

Here we can see $ sign but to make Price column categorical we have to remove it and type cast it to float64.

In [142]:
play_store_df['Price'] = play_store_df['Price'].apply(lambda x: float(str(x).replace('$', '')) if '$' in str(x) else float(x))

In [143]:
# Checking for the rows where Price is 0 and Type is Paid.
play_store_df[(play_store_df["Price"]==0) & (play_store_df["Type"]=="Paid")]

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content_Rating, Genres, Last_Updated, Current_Ver, Android_Ver]
Index: []

As we do not have any rows we will fill the single Type record with Nan value manually.

In [144]:
play_store_df["Type"].fillna('Free',inplace=True)

In [145]:
# Type casting the Last_updated column into Date Time Format  
from datetime import datetime
play_store_df["Last_Updated"]=play_store_df["Last_Updated"].apply(lambda x: datetime.strptime(x, "%B %d, %Y"))

In [146]:
print(list(play_store_df.Current_Ver.unique())[10:40])

['1.0.15', '3.8', '1.2.3', nan, '3.1', '2.2.5', '5.5.4', '4.0', '2.2.6.2', '1.1.3', '1.5', '1.0.8', '1.03', '6.0', '6.7.12.2018', '1.2', '2.20', '1.1.0', '1.6', '2.1', '1.0.9', '1.3', '1', '2.0.1', '1.46', '1.6.1', '11.0', '3.0', '1.7.1', '2.5.1']


List of Current_ver unique values we can't fill the nan values in this column because Apps are having different values in versions too not in particular sequential increasing order.

In [147]:
current_version_nan=list(play_store_df[play_store_df["Current_Ver"].isnull()]["App"])
for i in range(len(current_version_nan)):
  print(play_store_df[play_store_df['App']==current_version_nan[i]])

                                App        Category  ...  Current_Ver  Android_Ver
15  Learn To Draw Kawaii Characters  ART_AND_DESIGN  ...          NaN   4.2 and up

[1 rows x 13 columns]
                       App            Category  ...  Current_Ver  Android_Ver
1354  Market Update Helper  LIBRARIES_AND_DEMO  ...          NaN   1.5 and up

[1 rows x 13 columns]
                         App Category  ...  Current_Ver  Android_Ver
5863  Virtual DJ Sound Mixer    TOOLS  ...          NaN   4.0 and up

[1 rows x 13 columns]
            App Category    Rating  ...  Last_Updated  Current_Ver Android_Ver
6337  BT Master   FAMILY  4.191153  ...    2016-11-06          NaN  1.6 and up

[1 rows x 13 columns]
              App Category  Rating  ...  Last_Updated  Current_Ver Android_Ver
6864  Dots puzzle   FAMILY     4.0  ...    2018-04-18          NaN  4.0 and up

[1 rows x 13 columns]
                  App Category    Rating  ...  Last_Updated  Current_Ver Android_Ver
6938  Calculate My IQ   

Only 8 values are there so we will remove as this data is less than .001% of data.

In [148]:
# Checking null values in Android Version
play_store_df[play_store_df.Android_Ver.isnull()]

App         Category  ...  Current_Ver  Android_Ver
3998  [substratum] Vacuum: P  PERSONALIZATION  ...          4.4          NaN
4035    Pi Dark [substratum]  PERSONALIZATION  ...          1.1          NaN

[2 rows x 13 columns]

Only 2 null values.

In [149]:
play_store_df[play_store_df.App=="Pi Dark [substratum]"]

App         Category  ...  Current_Ver  Android_Ver
4035  Pi Dark [substratum]  PERSONALIZATION  ...          1.1          NaN

[1 rows x 13 columns]

In [150]:
play_store_df[play_store_df.App=="[substratum] Vacuum: P"]

App         Category  ...  Current_Ver  Android_Ver
3998  [substratum] Vacuum: P  PERSONALIZATION  ...          4.4          NaN

[1 rows x 13 columns]

We don't have different versions of these application so we will fill them with most used android version in that particular Year manually.

In [151]:
# Making a new column named Year
play_store_df["Year"]=play_store_df.Last_Updated.dt.year

In [152]:
play_store_df[play_store_df.Year==2018]['Android_Ver'].value_counts()

4.1 and up            1876
Varies with device    1056
4.0.3 and up          1013
4.0 and up             774
4.4 and up             748
5.0 and up             493
4.2 and up             294
2.3 and up             182
4.3 and up             175
3.0 and up              61
2.3.3 and up            58
6.0 and up              45
7.0 and up              39
2.1 and up              27
2.2 and up              24
5.1 and up              16
3.2 and up              11
1.6 and up              10
4.4W and up              9
8.0 and up               6
2.0 and up               5
7.1 and up               2
4.0.3 - 7.1.1            2
1.0 and up               1
2.0.1 and up             1
5.0 - 8.0                1
5.0 - 7.1.1              1
1.5 and up               1
7.0 - 7.1.1              1
Name: Android_Ver, dtype: int64

Android version 4.0 and above is mostly used in Year 2018 so we will fill null values with "4.1 and up".

In [153]:
play_store_df['Android_Ver'].fillna("4.1 and up", inplace=True)

In [154]:
# Dropping all the records where null values are there 
play_store_df.dropna(inplace=True)

In [155]:
play_store_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10349 entries, 0 to 10356
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10349 non-null  object        
 1   Category        10349 non-null  object        
 2   Rating          10349 non-null  float64       
 3   Reviews         10349 non-null  int64         
 4   Size            10349 non-null  float64       
 5   Installs        10349 non-null  object        
 6   Type            10349 non-null  object        
 7   Price           10349 non-null  float64       
 8   Content_Rating  10349 non-null  object        
 9   Genres          10349 non-null  object        
 10  Last_Updated    10349 non-null  datetime64[ns]
 11  Current_Ver     10349 non-null  object        
 12  Android_Ver     10349 non-null  object        
 13  Year            10349 non-null  int64         
dtypes: datetime64[ns](1), float64(3), int64(2), object(8)


No null values are there in our Dataframe.

In [156]:
play_store_df.describe(include="all")

App Category  ...  Android_Ver          Year
count    10349    10349  ...        10349  10349.000000
unique    9651       33  ...           33           NaN
top     ROBLOX   FAMILY  ...   4.1 and up           NaN
freq         9     1940  ...         2381           NaN
first      NaN      NaN  ...          NaN           NaN
last       NaN      NaN  ...          NaN           NaN
mean       NaN      NaN  ...          NaN   2017.383612
std        NaN      NaN  ...          NaN      1.112144
min        NaN      NaN  ...          NaN   2010.000000
25%        NaN      NaN  ...          NaN   2017.000000
50%        NaN      NaN  ...          NaN   2018.000000
75%        NaN      NaN  ...          NaN   2018.000000
max        NaN      NaN  ...          NaN   2018.000000

[13 rows x 14 columns]

### 1.3.2 Cleaning the user reviews dataframe

In [157]:
# Cleaning user reviews dataframe
user_reviews_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 5 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   App                     64295 non-null  object 
 1   Translated_Review       37427 non-null  object 
 2   Sentiment               37432 non-null  object 
 3   Sentiment_Polarity      37432 non-null  float64
 4   Sentiment_Subjectivity  37432 non-null  float64
dtypes: float64(2), object(3)
memory usage: 2.5+ MB


In [158]:
user_reviews_df.isnull().sum()

App                           0
Translated_Review         26868
Sentiment                 26863
Sentiment_Polarity        26863
Sentiment_Subjectivity    26863
dtype: int64

We will check for the null values wether they have common index or not and removing those null values.

In [159]:
sentiment_isnull_index=user_reviews_df[user_reviews_df.Sentiment.isnull()].index
sentiment_polarity_isnull_index=user_reviews_df[user_reviews_df.Sentiment_Polarity.isnull()].index
sentiment_subjectivity_isnull_index=user_reviews_df[user_reviews_df.Sentiment_Subjectivity.isnull()].index

In [160]:
if sentiment_isnull_index.all() == sentiment_polarity_isnull_index.all():
  print("All indexs are common in sentiment_isnull_index and sentiment_polarity_isnull_index")
if sentiment_isnull_index.all() == sentiment_subjectivity_isnull_index.all():
  print("All indexs are common in sentiment_isnull_index and sentiment_subjectivity_isnull_index")

All indexs are common in sentiment_isnull_index and sentiment_polarity_isnull_index
All indexs are common in sentiment_isnull_index and sentiment_subjectivity_isnull_index


In [161]:
user_reviews_df.drop(sentiment_subjectivity_isnull_index, inplace=True)

In [162]:
user_reviews_df.describe(include="all")

App  ... Sentiment_Subjectivity
count        37432  ...           37432.000000
unique         865  ...                    NaN
top     Bowmasters  ...                    NaN
freq           312  ...                    NaN
mean           NaN  ...               0.492704
std            NaN  ...               0.259949
min            NaN  ...               0.000000
25%            NaN  ...               0.357143
50%            NaN  ...               0.514286
75%            NaN  ...               0.650000
max            NaN  ...               1.000000

[11 rows x 5 columns]

Dropped the common indexes.
Now our both the data frames are cleaned handling of NaN/ Null/ Missing Values is completed.

# 1.4 Extraction of features information
Here all the features with the information they are providing present in our original dataframe is there.

### Play Store dataset contains following columns and their information:
1. App: Application name
2. Category: Category to which the app belongs (34 available)
3. Rating: Rating of the app (0 - 5)
4. Reviews: Number of reviews of the app available (Discrete)
5. Size: Size of the app in MB (Discrete)
6. Installs: Number of times the app has installed (Multi categorical)
7. Type: Free or Paid
8. Price: Price of the app
9. Content Rating: Whether it is for adults, teenagers etc. based on age(6 available)
10. Generes: Genere of the app
11. Last Updated: Last Update date
12. Current Ver: Current version of the application in playstore
13. Android Ver: Android version of the application

### User Reviews dataset contains following columns and their information:
1. App: Application name (duplicates present depending upon number of reviews present).
2. Translated_Review: Translated Reviews of that application.
3. Sentiment: Sentiment of the Review whether positive, negative or neutral.
4. Sentiment_Polarity: Polarity of the sentiment -ve vs +ve (-1 to +1).
5. Sentiment_Subjectivity: Sentiment is near to application subject subjective vs objective(0 - 1).

### Adding description about the sentiment analysis
**A sentiment is defined by its semantic orientation and the intensity of each word in the sentence.** This requires a pre-defined dictionary classifying negative and positive words. Generally, a text message will be represented by bag of words. After assigning individual scores to all the words, final sentiment is calculated by some pooling operation like taking an average of all the sentiments.

TextBlob returns polarity and subjectivity of a sentence. Polarity lies between [-1,1], -1 defines a negative sentiment and 1 defines a positive sentiment. Negation words reverse the polarity. TextBlob has semantic labels that help with fine-grained analysis. 

**Subjectivity quantifies the amount of personal opinion and factual information contained in the text. The higher subjectivity means that the text contains personal opinion rather than factual information.** TextBlob has one more parameter — intensity. TextBlob calculates subjectivity by looking at the ‘intensity’. Intensity determines if a word modifies the next word. For English, adverbs are used as modifiers (‘very good’).

For example: We calculated polarity and subjectivity for “I do not like this example at all, it is too boring”. For this particular example, polarity = -1 and subjectivity is 1, which is fair.

However, for the sentence “This was a helpful example but I would prefer another one”. It returns 0.0 for both subjectivity and polarity which is not the finest answer we’d expect.

In [163]:
print("Play store dataframe shape:",play_store_df.shape,"\nUser reviews dataframe shape:",user_reviews_df.shape)

Play store dataframe shape: (10349, 14) 
User reviews dataframe shape: (37432, 5)


# 1.5 Exploratory Data Analysis (EDA)

Trying to get some conclusion from data, correlation, trends which make sense and accomplishing milestones given in the problem statement by using visualization.

For each visualization we will create a new dataframe.


In [164]:
df1=play_store_df[["Year","Category"]].groupby(["Year","Category"],as_index=False).size()

In [165]:
fig = px.pie(df1, values='size', names='Category', hole = 0.4, title='Number of Apps present in dataframe according to Category')
fig.add_annotation(text = '2010-2018', font = dict(size=20, color='black'), showarrow = False)
fig.show()

Here we are able to see percentage of apps in each category.

In [166]:
# Making a new column name Month
play_store_df["Month"]=play_store_df.Last_Updated.dt.month

In [167]:
play_store_df[['Month','Year']]

Month  Year
0          1  2018
1          1  2018
2          8  2018
3          6  2018
4          6  2018
...      ...   ...
10352      7  2017
10353      7  2018
10354      1  2017
10355      1  2015
10356      7  2018

[10349 rows x 2 columns]

In [168]:
# Converting numerical month to textual months
import calendar
play_store_df['Month'] = play_store_df['Month'].apply(lambda x: calendar.month_name[x])

In [169]:
play_store_df[['Month','Year']]

Month  Year
0      January  2018
1      January  2018
2       August  2018
3         June  2018
4         June  2018
...        ...   ...
10352     July  2017
10353     July  2018
10354  January  2017
10355  January  2015
10356     July  2018

[10349 rows x 2 columns]

In [170]:
df2=play_store_df[["Category","Year","Month"]].groupby(["Year","Month","Category"], as_index=False).size()
df2.rename(columns={"size":"NumberOfApps"},inplace=True)

In [171]:
df2.head()

Year     Month   Category  NumberOfApps
0  2010       May     FAMILY             1
1  2011     April     FAMILY             1
2  2011     April       GAME             2
3  2011  December       GAME             1
4  2011  December  LIFESTYLE             1

In [172]:
month=["January", "February", "March", "April", "May", "June", 
      "July", "August", "September", "October", "November", "December"]

In [173]:
fig = px.bar(df2[df2['Year']==2018], x = 'Month', y = "NumberOfApps", color="Category", title='2018 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2017], x = 'Month', y = "NumberOfApps", color="Category", title='2017 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2016], x = 'Month', y = "NumberOfApps", color="Category", title='2016 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2015], x = 'Month', y = "NumberOfApps", color="Category", title='2015 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2014], x = 'Month', y = "NumberOfApps", color="Category", title='2014 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2013], x = 'Month', y = "NumberOfApps", color="Category", title='2013 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2012], x = 'Month', y = "NumberOfApps", color="Category", title='2012 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2011], x = 'Month', y = "NumberOfApps", color="Category", title='2011 Monthly Category Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df2[df2['Year']==2010], x = 'Month', y = "NumberOfApps", color="Category", title='2010 Monthly Category Trend', category_orders={"Month": month})
fig.show()

Here we can see monthly trend of category of applications.
We can find out in which month which type of applications are mostly released in each year(2010-2018).

In [174]:
df3=play_store_df[["Installs","Last_Updated","Year","Month","Category"]].groupby(["Year","Month","Installs","Category","Last_Updated"], as_index=False).size()
df3

Year     Month      Installs          Category Last_Updated  size
0     2010       May      100,000+            FAMILY   2010-05-21     1
1     2011     April    5,000,000+              GAME   2011-04-11     1
2     2011     April       50,000+            FAMILY   2011-04-18     1
3     2011     April       50,000+              GAME   2011-04-16     1
4     2011  December      100,000+         LIFESTYLE   2011-12-03     1
...    ...       ...           ...               ...          ...   ...
8421  2018       May      500,000+  TRAVEL_AND_LOCAL   2018-05-23     1
8422  2018       May      500,000+  TRAVEL_AND_LOCAL   2018-05-29     1
8423  2018       May  500,000,000+            FAMILY   2018-05-25     1
8424  2018       May  500,000,000+              GAME   2018-05-25     3
8425  2018       May  500,000,000+      PRODUCTIVITY   2018-05-23     1

[8426 rows x 6 columns]

In [175]:
# Here we are converting our Installs into numerical column
df3['Installs'] = df3['Installs'].apply(lambda x: str(x).replace('+','') if '+' in str(x) else x)
df3['Installs'] = df3['Installs'].apply(lambda x: str(x).replace(',', '') if ',' in str(x) else x)
df3['Installs']=df3['Installs'].astype('int')

In [176]:
fig = px.bar(df3[df3['Year']==2018], x = "Month", y = "Installs", color="Category", title='2018 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2017], x = 'Month', y = "Installs", color="Category", title='2017 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2016], x = 'Month', y = "Installs", color="Category", title='2016 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2015], x = 'Month', y = "Installs", color="Category", title='2015 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2014], x = 'Month', y = "Installs", color="Category", title='2014 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2013], x = 'Month', y = "Installs", color="Category", title='2013 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2012], x = 'Month', y = "Installs", color="Category", title='2012 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2011], x = 'Month', y = "Installs", color="Category", title='2011 Monthly installation Trend', category_orders={"Month": month})
fig.show()
fig = px.bar(df3[df3['Year']==2010], x = 'Month', y = "Installs", color="Category", title='2010 Monthly installation Trend', category_orders={"Month": month})
fig.show()

Here we are getting monthly trend of each category installed in each year which can provide us the installation details according festivals, holiday or political shift etc.

To get it in more better way we can compare month wise for 2018 and 2017 in the next graphs.

In [ ]:
df4=df3[["Year","Month","Installs","Category"]].groupby(["Year","Month","Category"],as_index=False).sum()

In [178]:
fig = px.bar(df4[df4['Year']==2018], x = "Category", y = "Installs", animation_frame="Month",text="Category", animation_group="Category", title='2018 Monthly installation Trend', category_orders={"Month": month})
fig.update_traces(textposition='outside')
fig.update_xaxes(visible=False)
fig.show()

In [179]:
fig = px.bar(df4[df4['Year']==2017], x = "Category", y = "Installs", animation_frame="Month", animation_group="Category", text="Category", title='2017 Monthly installation Trend', category_orders={"Month": month})
fig.update_traces(textposition='outside')
fig.update_xaxes(visible=False)
fig.show()

Monthly trend of installations of different categories.

In [ ]:
df5=play_store_df[["App","Year","Category","Type","Installs","Price"]].groupby(["Year","Category","Installs","Type","Price","App"], as_index=False).size()


In [181]:
fig = px.scatter(df5, x = "Year", y = "Price", color="Category", symbol="Category", title='Price of different Apps of different categories')
fig.show()

Here we have the different prices of the Apps of different categories yearly. We have multiple Apps also 